File: ui/backend/schemas/approval.py
"""Pydantic schemas for Approval Queue API.

Provides data validation schemas for the approval queue endpoints.
Uses Pydantic v2 for validation and serialization.

Schemas:
    - ApprovalQueueItemSchema: Individual approval queue item
    - ComplianceCheckSchema: Individual compliance check result
    - QualityBreakdownSchema: Quality score breakdown by factor
    - ApprovalQueueResponse: Paginated response for approval queue
"""

from datetime import datetime
from enum import IntEnum, Enum
from typing import Optional

from pydantic import BaseModel, Field


class SourcePriority(IntEnum):
    """Source-based priority for approval queue ordering.

    Lower values indicate higher priority (more urgent).

    Values:
        TRENDING: Time-sensitive trending content (highest priority)
        SCHEDULED: Content approaching deadline
        EVERGREEN: Flexible timing content
        RESEARCH: Research-based content (lowest urgency)
    """

    TRENDING = 1
    SCHEDULED = 2
    EVERGREEN = 3
    RESEARCH = 4


class ComplianceStatus(str, Enum):
    """Content compliance status.

    Indicates the result of EU Health Claims Regulation (EC 1924/2006)
    validation performed on content.

    Values:
        COMPLIANT: Content passes compliance checks
        WARNING: Borderline content, needs review
        REJECTED: Content contains prohibited claims
    """

    COMPLIANT = "COMPLIANT"
    WARNING = "WARNING"
    REJECTED = "REJECTED"


class QualityColor(str, Enum):
    """Quality score color coding.

    Used for visual indication of content quality.

    Values:
        GREEN: High quality (score >= 8)
        YELLOW: Medium quality (score >= 5 and < 8)
        RED: Low quality (score < 5)
    """

    GREEN = "green"
    YELLOW = "yellow"
    RED = "red"


def get_quality_color(score: float) -> QualityColor:
    """Calculate quality color from score.

    Args:
        score: Quality score (0-10)

    Returns:
        QualityColor based on score threshold
    """
    if score >= 8:
        return QualityColor.GREEN
    if score >= 5:
        return QualityColor.YELLOW
    return QualityColor.RED


class ComplianceCheckSchema(BaseModel):
    """Individual compliance check result.

    Represents a single compliance check performed on content,
    including the phrase checked, result status, and explanation.
    """

    phrase: str = Field(
        ...,
        description="The phrase that was checked for compliance",
    )
    status: str = Field(
        ...,
        description="Check result: 'prohibited', 'borderline', or 'permitted'",
    )
    explanation: str = Field(
        ...,
        description="Human-readable explanation of the check result",
    )
    regulation_reference: Optional[str] = Field(
        default=None,
        description="EU regulation reference if applicable",
    )

    model_config = {
        "from_attributes": True,
    }


class QualityBreakdownSchema(BaseModel):
    """Quality score breakdown by factor.

    Each factor contributes to the overall quality score with
    specified weights.
    """

    compliance_score: float = Field(
        ...,
        ge=0,
        le=10,
        description="Compliance factor score (25% weight)",
    )
    brand_voice_score: float = Field(
        ...,
        ge=0,
        le=10,
        description="Brand voice alignment score (20% weight)",
    )
    visual_quality_score: float = Field(
        ...,
        ge=0,
        le=10,
        description="Visual quality score (15% weight)",
    )
    platform_optimization_score: float = Field(
        ...,
        ge=0,
        le=10,
        description="Platform optimization score (15% weight)",
    )
    engagement_prediction_score: float = Field(
        ...,
        ge=0,
        le=10,
        description="Engagement prediction score (15% weight)",
    )
    authenticity_score: float = Field(
        ...,
        ge=0,
        le=10,
        description="Authenticity score (10% weight)",
    )

    model_config = {
        "from_attributes": True,
    }


class ApprovalQueueItemSchema(BaseModel):
    """Schema for approval queue item.

    Contains all fields needed for queue display and detail view.
    """

    id: str = Field(
        ...,
        description="Unique identifier for the approval item",
    )
    thumbnail_url: str = Field(
        ...,
        description="URL to thumbnail image (200x200)",
    )
    caption_excerpt: str = Field(
        ...,
        max_length=100,
        description="First 100 characters of caption",
    )
    full_caption: str = Field(
        ...,
        description="Complete caption text",
    )
    quality_score: float = Field(
        ...,
        ge=0,
        le=10,
        description="Overall quality score (0-10)",
    )
    quality_color: QualityColor = Field(
        ...,
        description="Color coding based on quality score",
    )
    compliance_status: ComplianceStatus = Field(
        ...,
        description="Content compliance status",
    )
    would_auto_publish: bool = Field(
        ...,
        description="Whether content meets auto-publish criteria",
    )
    suggested_publish_time: Optional[datetime] = Field(
        default=None,
        description="Suggested time for publishing",
    )
    source_type: str = Field(
        ...,
        description="Content source type (instagram_post, b2b_email, etc.)",
    )
    source_priority: SourcePriority = Field(
        ...,
        description="Source-based priority for ordering",
    )
    hashtags: list[str] = Field(
        default_factory=list,
        description="List of hashtags for the content",
    )
    compliance_details: Optional[list[ComplianceCheckSchema]] = Field(
        default=None,
        description="Detailed compliance check results",
    )
    quality_breakdown: Optional[QualityBreakdownSchema] = Field(
        default=None,
        description="Quality score breakdown by factor",
    )
    created_at: datetime = Field(
        ...,
        description="When the content was created",
    )
    # Story 4-2: Edit support fields
    original_caption: Optional[str] = Field(
        default=None,
        description="Original caption before any edits (for revert)",
    )
    rewrite_suggestions: Optional[list[dict]] = Field(
        default=None,
        description="AI-suggested rewrites for compliance/quality improvement",
    )
    status: Optional[str] = Field(
        default=None,
        description="Current approval status (pending, approved, rejected, scheduled, publishing, published, publish_failed)",
    )
    # Story 4-5: Instagram publishing fields
    instagram_post_id: Optional[str] = Field(
        default=None,
        description="Instagram post ID after successful publish",
    )
    instagram_permalink: Optional[str] = Field(
        default=None,
        description="Direct link to Instagram post",
    )
    published_at: Optional[datetime] = Field(
        default=None,
        description="Timestamp when post was published to Instagram",
    )
    publish_error: Optional[str] = Field(
        default=None,
        description="Error message if publish failed",
    )
    publish_attempts: int = Field(
        default=0,
        ge=0,
        description="Number of publish attempts made",
    )

    model_config = {
        "from_attributes": True,
    }


class ApprovalQueueResponse(BaseModel):
    """Response schema for approval queue endpoint.

    Includes paginated items list with cursor-based pagination.
    """

    items: list[ApprovalQueueItemSchema] = Field(
        ...,
        description="List of approval queue items",
    )
    total_count: int = Field(
        ...,
        ge=0,
        description="Total number of items in queue",
    )
    next_cursor: Optional[str] = Field(
        default=None,
        description="Cursor for next page of results",
    )
    has_more: bool = Field(
        ...,
        description="Whether more items exist beyond this page",
    )

    model_config = {
        "from_attributes": True,
    }


__all__ = [
    "SourcePriority",
    "ComplianceStatus",
    "QualityColor",
    "get_quality_color",
    "ComplianceCheckSchema",
    "QualityBreakdownSchema",
    "ApprovalQueueItemSchema",
    "ApprovalQueueResponse",
]

File: ui/frontend-react/src/types/approval.ts
/**
 * TypeScript types for Approval Queue.
 *
 * Mirrors the backend Pydantic schemas for type-safe API communication.
 */

/**
 * Source-based priority for approval queue ordering.
 * Lower values indicate higher priority (more urgent).
 */
export enum SourcePriority {
  TRENDING = 1,
  SCHEDULED = 2,
  EVERGREEN = 3,
  RESEARCH = 4,
}

/**
 * Content compliance status.
 */
export enum ComplianceStatus {
  COMPLIANT = "COMPLIANT",
  WARNING = "WARNING",
  REJECTED = "REJECTED",
}

/**
 * Quality score color coding.
 */
export type QualityColor = "green" | "yellow" | "red";

/**
 * Calculate quality color from score.
 *
 * @param score - Quality score (0-10)
 * @returns Color based on score threshold
 */
export function getQualityColor(score: number): QualityColor {
  if (score >= 8) return "green";
  if (score >= 5) return "yellow";
  return "red";
}

/**
 * Individual compliance check result.
 */
export interface ComplianceCheck {
  phrase: string;
  status: "prohibited" | "borderline" | "permitted";
  explanation: string;
  regulation_reference?: string;
}

/**
 * Quality score breakdown by factor.
 */
export interface QualityBreakdown {
  compliance_score: number;
  brand_voice_score: number;
  visual_quality_score: number;
  platform_optimization_score: number;
  engagement_prediction_score: number;
  authenticity_score: number;
}

/**
 * Quality breakdown factor with weight information.
 */
export interface QualityFactor {
  key: keyof QualityBreakdown;
  label: string;
  weight: number;
  score: number;
}

/**
 * Quality factors with their weights.
 */
export const QUALITY_FACTORS: Array<{ key: keyof QualityBreakdown; label: string; weight: number }> = [
  { key: "compliance_score", label: "Compliance", weight: 0.25 },
  { key: "brand_voice_score", label: "Brand Voice", weight: 0.20 },
  { key: "visual_quality_score", label: "Visual Quality", weight: 0.15 },
  { key: "platform_optimization_score", label: "Platform", weight: 0.15 },
  { key: "engagement_prediction_score", label: "Engagement", weight: 0.15 },
  { key: "authenticity_score", label: "Authenticity", weight: 0.10 },
];

/**
 * Approval queue item from API.
 */
export interface ApprovalQueueItem {
  id: string;
  thumbnail_url: string;
  caption_excerpt: string;
  full_caption: string;
  quality_score: number;
  quality_color: QualityColor;
  compliance_status: ComplianceStatus;
  would_auto_publish: boolean;
  suggested_publish_time: string | null;
  source_type: string;
  source_priority: SourcePriority;
  hashtags: string[];
  compliance_details?: ComplianceCheck[];
  quality_breakdown?: QualityBreakdown;
  created_at: string;
  // Story 4-2: Edit/revert support
  original_caption?: string | null;
  rewrite_suggestions?: RewriteSuggestion[];
  // Approval/rejection tracking
  status?: ApprovalStatus;
  approved_at?: string | null;
  approved_by?: string | null;
  rejection_reason?: RejectReason | null;
  rejection_text?: string | null;
}

/**
 * Paginated response from approval queue API.
 */
export interface ApprovalQueueResponse {
  items: ApprovalQueueItem[];
  total_count: number;
  next_cursor: string | null;
  has_more: boolean;
}

/**
 * Source type display configuration.
 */
export interface SourceTypeConfig {
  label: string;
  icon: string;
  color: string;
}

/**
 * Source type configurations.
 */
export const SOURCE_TYPES: Record<string, SourceTypeConfig> = {
  instagram_post: {
    label: "Instagram Post",
    icon: "instagram",
    color: "bg-pink-100 text-pink-800",
  },
  b2b_email: {
    label: "B2B Email",
    icon: "mail",
    color: "bg-blue-100 text-blue-800",
  },
  story: {
    label: "Story",
    icon: "clock",
    color: "bg-purple-100 text-purple-800",
  },
  reel: {
    label: "Reel",
    icon: "video",
    color: "bg-orange-100 text-orange-800",
  },
};

/**
 * Approval item status for workflow states.
 */
export enum ApprovalStatus {
  PENDING = "pending",
  APPROVED = "approved",
  REJECTED = "rejected",
  SCHEDULED = "scheduled",
  PUBLISHED = "published",
  PUBLISH_FAILED = "publish_failed",
}

/**
 * Rejection reason enum for predefined options.
 */
export enum RejectReason {
  COMPLIANCE_ISSUE = "compliance_issue",
  BRAND_VOICE_MISMATCH = "brand_voice_mismatch",
  LOW_QUALITY = "low_quality",
  IRRELEVANT_CONTENT = "irrelevant_content",
  DUPLICATE_CONTENT = "duplicate_content",
  OTHER = "other",
}

/**
 * Rejection reason labels for UI display.
 */
export const REJECT_REASON_LABELS: Record<RejectReason, string> = {
  [RejectReason.COMPLIANCE_ISSUE]: "Contains prohibited claims",
  [RejectReason.BRAND_VOICE_MISMATCH]: "Doesn't match DAWO tone",
  [RejectReason.LOW_QUALITY]: "Quality score too low",
  [RejectReason.IRRELEVANT_CONTENT]: "Topic not suitable",
  [RejectReason.DUPLICATE_CONTENT]: "Similar post already exists",
  [RejectReason.OTHER]: "Other reason",
};

/**
 * Schema for approve action request.
 */
export interface ApproveActionRequest {
  scheduled_publish_time?: string | null;
}

/**
 * Schema for reject action request.
 */
export interface RejectActionRequest {
  reason: RejectReason;
  reason_text?: string | null;
}

/**
 * Schema for edit action request.
 */
export interface EditActionRequest {
  caption: string;
  hashtags?: string[] | null;
}

/**
 * Revalidation result from compliance/quality checks.
 */
export interface RevalidationResult {
  compliance_status: ComplianceStatus;
  compliance_details?: ComplianceCheck[] | null;
  quality_score: number;
  quality_breakdown?: QualityBreakdown | null;
  rewrite_suggestions?: RewriteSuggestion[] | null;
}

/**
 * AI rewrite suggestion.
 */
export interface RewriteSuggestion {
  id: string;
  original_text: string;
  suggested_text: string;
  reason: string;
  type: "compliance" | "brand_voice" | "quality";
}

/**
 * Standard response for approval actions.
 */
export interface ApprovalActionResponse {
  success: boolean;
  message: string;
  item_id: string;
  new_status: ApprovalStatus;
  revalidation?: RevalidationResult | null;
}

/**
 * Edit history entry.
 */
export interface EditHistoryEntry {
  id: string;
  previous_caption: string;
  new_caption: string;
  edited_at: string;
  editor: string;
}

/**
 * Action type for approval actions hook.
 */
export type ApprovalActionType = "approve" | "reject" | "edit" | "revalidate";
